### DISCUSSION TITLE
How does Woodpecker CI deals with pod deletion?

### DISCUSSION THREAD

**Discussion by lara-clink** [woodpecker-ci/woodpecker#3835] [Ask a Question] [answered]
I've been experiencing issues with the delete pod operation in my Woodpecker CI pipelines. Specifically, I've encountered the following error message:
`Oh no, we got some errors!
Delete "https://100.64.0.1:443/api/v1/namespaces/woodpecker-pipelines/pods/wp-01hzf6tbz7g17adzarxnhpkd5z": http2: client connection lost`
It seems this issue might be related to an agent restart due to a memory leak. I have a few questions regarding the reliability of the delete pod operation and the order of operations in the pipeline:

1. Is the delete pod operation unreliable?

2. Does this operation require retries to handle transient errors like the one mentioned above?

3. Is the order of operations unreliable?

For example, should a step in the pipeline fail if the pod fails to be deleted? Or should there be some form of error handling to ensure the pipeline continues running despite pod deletion failures?
I'm looking to understand if there are existing mechanisms within Woodpecker CI to address these issues or if there are recommended practices for enhancing the reliability of these operations.

Thank you for your assistance!

**Comment by zc-devs** [comment, marked as answer]
> Is the delete pod operation unreliable?

Define `reliable` :)

> Delete "https://100.64.0.1:443/api/v1/namespaces/woodpecker-pipelines/pods/wp-01hzf6tbz7g17adzarxnhpkd5z": http2: client connection lost

This looks like Agent cannot call Kubernetes API.

> issue might be related to an agent restart due

That might happen. I used to have abandoned PVC/PV after Agent crash.

> Does this operation require retries to handle transient errors like the one mentioned above?

Would be great, if it is not implemented yet.

> should a step in the ~~pipeline~~ workflow fail if the pod fails to be deleted?

This one, I believe.
Edit: ⬆️ _by default_, however there is [on failure](https://woodpecker-ci.org/docs/usage/workflow-syntax#failure) condition.

---

Almost all is [there](https://github.com/woodpecker-ci/woodpecker/tree/main/pipeline/backend/kubernetes), start from [kubernetes.go](https://github.com/woodpecker-ci/woodpecker/blob/main/pipeline/backend/kubernetes/kubernetes.go).
